# Module

## 모듈이란?
 - 함수는 하나의 파일 내에서 손쉽게 코드를 재사용하는 방법
 - 그렇다면 여러 함수들뿐만 아니라 변수, 클래스들을 한꺼번에 불러들여 재사용하는 방법은?
 - 이럴때 모듈을 사용함

### 모듈 만들고 불러 보기

In [8]:
# mod1.py
def sum(a, b):
	return a + b

이 파일과 같은 위치에서 인터프리터를 실행하고 아래와 같이 실행해보자. 
<pre>>>> import mod1
>>> print(mod1.sum(3,4))
7</pre>

import는 미리 만들어져 있는 파이썬 모듈을 사용할 수 있게 해주는 명령어이다.
import의 사용법은 다음과 같다.
> import 모듈이름

In [10]:
# mod1.py 
def sum(a, b): 
    return a+b

def safe_sum(a, b): 
    if type(a) != type(b): 
        print("더할수 있는 것이 아닙니다.")
        return 
    else: 
        result = sum(a, b) 
    return result

if __name__ == "__main__":
	print(safe_sum('a', 1))
	print(safe_sum(1, 4))
	print(sum(10, 10.4))
	print(__name__)

더할수 있는 것이 아닙니다.
None
5
20.4
__main__


In [ ]:
# mod1.py
def sum(a, b):
	return a + b

def safe_sum(a, b): 
    if type(a) != type(b): 
        print("더할수 있는 것이 아닙니다.")
        return 
    else: 
        result = sum(a, b) 
    return result

또 다시 import를 한 후, 새로 추가한 함수를 호출할 수 있다.
<pre>
>>> import mod1
>>> print(mod1.safe_sum(3,4))
7
>>> print(mod1.safe_sum(1, 'a')
더할 수 있는 것이 아닙니다.
None
>>> 
</pre>

#### Note
위와 같이 그냥 mod1.py에 함수를 추가한뒤, import mod1을 실행한후, mode1.safe_sum(3,4)를 호출하면 아래와 같이 에러가 발생한다.
파이썬은 효율성의 이유로 하나의 인터프리터 세션에서 한번만 import 되도록 되어 있다고 한다. 따라서, 변경된 모듈을 인식하려면, 인터프리터를 재시작하거나, 아래와 같이 importlib 모듈을 사용하면된다.:)
```
>>> import mod1
>>> print(mod1.safe_sum(3,4))
Traceback (most recent call last):
  File "C:\Users\woojin\ml_venv\lib\site-packages\IPython\core\interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-61f00c3e6b60>", line 1, in <module>
    print(mod1.safe_sum(3,4))
AttributeError: module 'mod1' has no attribute 'safe_sum'
>>> import importlib
>>> importlib.reload(mod1)
Out[7]: <module 'mod1' from 'C:\\Users\\woojin\\PycharmProjects\\ai_study\\mod1.py'>
>>> print(mod1.safe_sum(3,4))
7
```

### 모듈 함수를 사용하는 또 다른 방법
 - 위와 같이 mod1.sum, mod1.safe_sum 과 같이 호출하는 방법은 매우 번거롭다.
 - 그냥 sum, safe_sum 과 같이 쓸수 없나?
 - 그럴때는 아래와 같은 형태로 사용한다.
> from 모듈이름 import 모듈함수
 - mod1에 적용해 보면 아래와 같다.
<pre>
>>> from mod1 import sum
>>> sum(3, 4)
7
</pre>
 - 그런데 위와 같이 하면 sum 함수만 사용가능하다. safe_sum도 사용하고 싶으면 어떻게 해야 할까?
 - 2가지 방법이 있다.
> from mod1 import sum, safe_sum
<br>
> from mod1 import *


<pre>if __name__ == "__main__": 의 의미?</pre>

In [ ]:
# mod1.py 
def sum(a, b): 
    return a+b

def safe_sum(a, b): 
    if type(a) != type(b): 
        print("더할수 있는 것이 아닙니다.")
        return 
    else: 
        result = sum(a, b) 
    return result 

print(safe_sum('a', 1))
print(safe_sum(1, 4))
print(sum(10, 10.4))

실행하면 다음과 같이 우리가 기대했던 결과가 나온다.
<pre>C:\python mod1.py
더할수 있는 것이 아닙니다.
None
5
20.4
</pre>

**그런데 문제는 우리가 이전과 같이 sum, safe_sum 함수만 사용하기 위해 import를 할 때 발생한다.**

```
>>> import mod1
더할수 있는 것이 아닙니다.
None
5
20.4
``` 

In [10]:
# mod1.py 
def sum(a, b): 
    return a+b

def safe_sum(a, b): 
    if type(a) != type(b): 
        print("더할수 있는 것이 아닙니다.")
        return 
    else: 
        result = sum(a, b) 
    return result

if __name__ == "__main__":
	print(safe_sum('a', 1))
	print(safe_sum(1, 4))
	print(sum(10, 10.4))
	print(__name__)

더할수 있는 것이 아닙니다.
None
5
20.4
__main__


즉, python mod1.py 로 실행했을때는 __name__ 의 값이 "__main__" 이 되어 print 문이 실행이 되었고
반대로 인터프리터나 다른 파일에서 import를 사용하여 모듈을 불러와서 사용할 때는 __name__ 값이 거짓이 되어 print 문이 실행이 되지 않음을 알 수 있다.